CrossSectionAnalysis/cpp/CrossSectionCompute.cxx

```diff
@@ -0,0 +1,340 @@
+          double tupleValues[3] = {0.0, 0.0, 0.0};
+          tupleValues[0] = bufferArray->GetTuple3(r)[0]; // Output table row index
+          tupleValues[1] = bufferArray->GetTuple3(r)[1]; // Cross-section area
+          tupleValues[2] = bufferArray->GetTuple3(r)[2]; // CE diameter
+          crossSectionAreaArray->SetValue((int) tupleValues[0], tupleValues[1]);
+          ceDiameterArray->SetValue((int) tupleValues[0], tupleValues[2]);
+      }
+  }
+  mtx.unlock();
+}
+
+/////////////////////////////////////////////////////////////////////////
+
+vtkCrossSectionComputeWorker::vtkCrossSectionComputeWorker()
+{
+}
+
+vtkCrossSectionComputeWorker::~vtkCrossSectionComputeWorker()
+{
+}
+
+void vtkCrossSectionComputeWorker::operator () (vtkMRMLNode * inputCenterlineNode,
+                             vtkPolyData * closedSurfacePolyData,
+                             vtkDoubleArray * bufferArray,
+                             unsigned int startPointIndex,
+                             unsigned int endPointIndex)
+{
+#if DEVTIME != 0
+    double startTime = GetTimeOfDay();
+#endif
+    for (unsigned int i = startPointIndex; i <= endPointIndex; i++)
+    {
+        // Get the contout polydata
+        vtkPolyData * polydata = ComputeCrossSectionPolydata(inputCenterlineNode, closedSurfacePolyData, i);
+        {
+            // Get the surface area and circular equivalent diameter
+            vtkNew<vtkMassProperties> crossSectionProperties;
+            crossSectionProperties->SetInputData(polydata);
+            const double crossSectionSurfaceArea = crossSectionProperties->GetSurfaceArea();
+            const double ceDiameter = (sqrt(crossSectionSurfaceArea / vtkMath::Pi())) * 2;
+            
+            bufferArray->InsertNextTuple3((double) i, crossSectionSurfaceArea, ceDiameter);
+        }
+        polydata->Delete();
+    }
+    
+    // Manage last point for an input centerline model
+    if (inputCenterlineNode->IsA("vtkMRMLModelNode"))
+    {
+        // Cross-section area is not calculated at last point for a model.
+        vtkMRMLModelNode * inputModel = vtkMRMLModelNode::SafeDownCast(inputCenterlineNode);
+        vtkPoints * modelPoints = inputModel->GetMesh()->GetPoints();
+        const unsigned int numberOfPoints = modelPoints->GetNumberOfPoints();
+        if (endPointIndex == (numberOfPoints - 1))
+        {
+            const unsigned int blockLastPointIndex = bufferArray->GetNumberOfTuples() - 1;
+            const unsigned int blockBeforeLastPointIndex = bufferArray->GetNumberOfTuples() - 1 - 1;
+            bufferArray->SetTuple3(blockLastPointIndex,
+                                   bufferArray->GetTuple3(blockLastPointIndex)[0],
+                                   bufferArray->GetTuple3(blockBeforeLastPointIndex)[1],
+                                   bufferArray->GetTuple3(blockBeforeLastPointIndex)[2]);
+        }
+    }
+    
+    // The centerline copy is not needed anymore.
+    inputCenterlineNode->Delete();
```

Comment:
Deleting a node triggers a lots of modifications. If there are observers in the main thread then the application will likely crash. The best would be not to pass nodes to the thread, but if you do then don't modify or delete those nodes in the thread.